stl/src/vector_algorithms.cpp

```diff
@@ -2009,6 +2027,79 @@ namespace {
         }
         return _Result;
     }
+
+    template <class _Traits, class _Ty>
+    __declspec(noalias) size_t
+        __stdcall _Mismatch(const void* const _First1, const void* const _First2, const size_t _Count) noexcept {
+        size_t _Result = 0;
+#ifndef _M_ARM64EC
+        const auto _First1_ch = static_cast<const char*>(_First1);
+        const auto _First2_ch = static_cast<const char*>(_First2);
+
+        if (_Use_avx2()) {
+            const size_t _Count_bytes          = _Count * sizeof(_Ty);
+            const size_t _Count_bytes_avx_full = _Count_bytes & ~size_t{0x1F};
+
+            for (; _Result != _Count_bytes_avx_full; _Result += 0x20) {
+                const __m256i _Elem1 = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(_First1_ch + _Result));
+                const __m256i _Elem2 = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(_First2_ch + _Result));
+                const auto _Bingo = ~static_cast<unsigned int>(_mm256_movemask_epi8(_Traits::_Cmp_avx(_Elem1, _Elem2)));
+                if (_Bingo != 0) {

```

Comment:
If you're expecting this loop to run long, perhaps you should swap this for if (!_mm256_testc_si256(traits::cmp_avx(), _mm256_set1_epi32(-1))), then put the bingo stuff inside the braces.

(If you expect it to run short, that would be a pessimization; the movemask is needed for the last iteration. Make sure to test it.)

Reply:
I'll measure.
Ideally I don't want to introduce vectorization threshold and favor smaller ranges too.

Reply:
There's no unambiguous winner. The below strange results are consistent.
I suspect some thin effect, like the exact loop alignment.
I'll keep the current version, it is just simpler.

current
```
C:\Project\STL>out\bench\benchmark-mismatch.exe --benchmark_min_time=2s
2024-03-23T07:58:04+02:00
Running out\bench\benchmark-mismatch.exe
Run on (12 X 2208 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
bm<uint8_t>/8/3              5.66 ns         5.65 ns    497777778
bm<uint8_t>/24/22            5.62 ns         5.61 ns    437073171
bm<uint8_t>/105/-1           10.6 ns         10.5 ns    259710145
bm<uint8_t>/4021/3056         111 ns          111 ns     25239437
bm<uint16_t>/8/3             4.81 ns         4.81 ns    597333333
bm<uint16_t>/24/22           6.34 ns         6.36 ns    484324324
bm<uint16_t>/105/-1          9.52 ns         9.52 ns    303728814
bm<uint16_t>/4021/3056        129 ns          129 ns     22123457
bm<uint32_t>/8/3             4.73 ns         4.71 ns    597333333
bm<uint32_t>/24/22           5.62 ns         5.62 ns    497777778
bm<uint32_t>/105/-1          13.5 ns         13.6 ns    154482759
bm<uint32_t>/4021/3056        250 ns          251 ns     10541176
bm<uint64_t>/8/3             4.96 ns         4.96 ns    617931034
bm<uint64_t>/24/22           7.83 ns         7.83 ns    381276596
bm<uint64_t>/105/-1          20.2 ns         20.2 ns    137846154
bm<uint64_t>/4021/3056        516 ns          515 ns      5430303
```

vptest
```
C:\Project\STL>out\bench\benchmark-mismatch.exe --benchmark_min_time=2s
2024-03-23T07:54:44+02:00
Running out\bench\benchmark-mismatch.exe
Run on (12 X 2208 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
bm<uint8_t>/8/3              5.74 ns         5.73 ns    471578947
bm<uint8_t>/24/22            5.75 ns         5.77 ns    484324324
bm<uint8_t>/105/-1           9.13 ns         9.10 ns    298666667
bm<uint8_t>/4021/3056        75.7 ns         75.8 ns     37333333
bm<uint16_t>/8/3             5.47 ns         5.46 ns    497777778
bm<uint16_t>/24/22           6.73 ns         6.72 ns    381276596
bm<uint16_t>/105/-1          9.93 ns         9.94 ns    284444444
bm<uint16_t>/4021/3056        134 ns          134 ns     21590361
bm<uint32_t>/8/3             4.58 ns         4.60 ns    617931034
bm<uint32_t>/24/22           5.84 ns         5.84 ns    484324324
bm<uint32_t>/105/-1          12.7 ns         12.7 ns    208372093
bm<uint32_t>/4021/3056        242 ns          242 ns     11946667
bm<uint64_t>/8/3             4.83 ns         4.84 ns    578064516
bm<uint64_t>/24/22           7.61 ns         7.60 ns    365714286
bm<uint64_t>/105/-1          20.1 ns         20.0 ns    128000000
bm<uint64_t>/4021/3056        552 ns          553 ns      4715789
```

like current, but with subtraction
```
C:\Project\STL>out\bench\benchmark-mismatch.exe --benchmark_min_time=2s
2024-03-23T08:11:11+02:00
Running out\bench\benchmark-mismatch.exe
Run on (12 X 2208 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
bm<uint8_t>/8/3              5.03 ns         5.04 ns    543030303
bm<uint8_t>/24/22            5.01 ns         4.99 ns    560000000
bm<uint8_t>/105/-1           9.33 ns         9.31 ns    303728814
bm<uint8_t>/4021/3056        91.1 ns         91.0 ns     31438596
bm<uint16_t>/8/3             5.20 ns         5.22 ns    560000000
bm<uint16_t>/24/22           6.16 ns         6.15 ns    426666667
bm<uint16_t>/105/-1          9.79 ns         9.75 ns    275692308
bm<uint16_t>/4021/3056        126 ns          127 ns     21590361
bm<uint32_t>/8/3             4.88 ns         4.88 ns    560000000
bm<uint32_t>/24/22           6.08 ns         6.09 ns    459487179
bm<uint32_t>/105/-1          15.4 ns         15.5 ns    182857143
bm<uint32_t>/4021/3056        308 ns          309 ns      8960000
bm<uint64_t>/8/3             5.04 ns         5.05 ns    578064516
bm<uint64_t>/24/22           9.53 ns         9.56 ns    308965517
bm<uint64_t>/105/-1          31.2 ns         31.3 ns     85333333
bm<uint64_t>/4021/3056        818 ns          818 ns      3381132
```

Reply:
Looks roughly as expected for me. vptest is mostly faster for long-running loops, but for short ones, reusing more instructions between main loop and the found-it branch reduces the constant factor. Also plenty of noise, which is also expected; modern cpu performance is more random than std::mt19937.

Yet another thing you could try is skip the cmpeq. xor elem1/elem2 instead, then _mm256_testz_si256 the result with itself. That saves having to fill a register with -1 in the prologue (or in the actual loop, if the compiler doesn't hoist it).

But that's at the cost of putting the cmpeq in the found-it branch, meaning one more instruction if the target is found early, so it'll probably just amplify the wins and losses you've already seen.

Probably not worth it; your choice if it's even worth checking. The bottleneck is probably memory bandwidth and not the computation, anyways.

Reply:
`vpxor`+`vptest`
```
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
bm<uint8_t>/8/3              6.10 ns         6.10 ns    448000000
bm<uint8_t>/24/22            6.22 ns         6.23 ns    426666667
bm<uint8_t>/105/-1           9.04 ns         9.04 ns    293770492
bm<uint8_t>/4021/3056        85.4 ns         85.3 ns     29866667
bm<uint16_t>/8/3             5.43 ns         5.42 ns    484324324
bm<uint16_t>/24/22           6.87 ns         6.86 ns    389565217
bm<uint16_t>/105/-1          9.70 ns         9.68 ns    289032258
bm<uint16_t>/4021/3056        139 ns          137 ns     20363636
bm<uint32_t>/8/3             5.13 ns         5.14 ns    578064516
bm<uint32_t>/24/22           5.92 ns         5.95 ns    459487179
bm<uint32_t>/105/-1          13.7 ns         13.7 ns    205977011
bm<uint32_t>/4021/3056        267 ns          265 ns      9955556
bm<uint64_t>/8/3             4.80 ns         4.80 ns    560000000
bm<uint64_t>/24/22           8.20 ns         8.20 ns    325818182
bm<uint64_t>/105/-1          23.4 ns         23.4 ns    119466667
bm<uint64_t>/4021/3056        665 ns          664 ns      4167442
```
Looks like a worse version of `vpcmpeqb`+`vptest`